lib/spack/spack/solver/asp.py

```diff
@@ -2840,6 +2797,87 @@ def reject_requirement_constraint(
         return False
 
 
+class CompilerParser:
+    def __init__(self, configuration) -> None:
+        self.compilers: Set[KnownCompiler] = set()
+        for c in all_compilers_in_config(configuration):
+            target = c.target if c.target != "any" else None
+            candidate = KnownCompiler(
+                spec=c.spec, os=c.operating_system, target=target, available=True, compiler_obj=c
+            )
+            if candidate in self.compilers:
+                warnings.warn(
+                    f"duplicate found for {c.spec} on {c.operating_system}/{c.target}. "
+                    f"Edit your compilers.yaml configuration to remove it."
+                )
+                continue
+
+            self.compilers.add(candidate)
+
+    def with_input_specs(self, input_specs: List["spack.spec.Spec"]) -> "CompilerParser":
+        strict = spack.concretize.Concretizer().check_for_compiler_existence
+        default_os = str(spack.platforms.host().default_os)
+        default_target = str(archspec.cpu.host().family)
+        for s in traverse.traverse_nodes(input_specs):
+            # we don't need to validate compilers for already-built specs
+            if s.concrete or not s.compiler:
+                continue
+
+            version = s.compiler.versions.concrete
+
+            if not version or any(item.spec.satisfies(s.compiler) for item in self.compilers):
```

Comment:
Isn't s.compiler the same object `id` almost always? Given that a compiler is almost always constant this sounds inefficient.

Reply:
Also `not version` should in principle be implied by `s.concrete`

Reply:
> satisfies subset semantics makes not sense for concrete specs anyways.

Is this based on a misunderstanding:
```python
for s in traverse.traverse_nodes(input_specs):
    # we don't need to validate compilers for already-built specs
    if s.concrete or not s.compiler:
        continue
```
? If `s` is concrete we won't reach that code, and move on to the next iteration. Note that this is run for input specs, not on reusable concrete specs.

Reply:
I'm also shortening the check in the other method.